Commissioning script

aninternetof opened the pull request:
This creates a new systemd service to run a commissioning python script. There is nothing in the script yet.

aninternetof commented on README.md:
```diff
@@ -17,38 +17,48 @@ Rovercode runs on a [Raspberry Pi 3](https://www.raspberrypi.org/products/raspbe
+The file will download as something like `rovercode_yourrovername.env`.
+Save this file to a flash drive.
 
 ### Rover Setup
```
When installing on the Rover, we only need to install Docker, then put the systemd service files and the scripts in the right places.

aninternetof commented on services/rovercode.service:
```diff
@@ -10,7 +10,8 @@ User=root
 ExecStartPre=-/usr/bin/docker rm rovercode
-ExecStart=-/usr/bin/docker run --name rovercode --privileged -v /dev:/dev -v /opt/rovercode:/var/rovercode rovercode
+ExecStartPre=-/usr/bin/docker pull rovercode/rovercode:latest
+ExecStart=-/usr/bin/docker run --name rovercode --privileged -v /dev:/dev -v /opt/rovercode/.env:/var/rovercode/.env rovercode
```
The only thing we need to map in is the `.env` file.

aninternetof commented on services/commissioning.service:
```diff
@@ -0,0 +1,18 @@
+User=root
+Group=root
+ExecStartPre=-/usr/bin/docker stop rovercode
+ExecStart=-/usr/bin/docker run --name rovercode-commissioning -v /opt/rovercode/scripts/commission.py:/var/rovercode/scripts/commission.py --entrypoint=/bin/bash rovercode -c 'python /var/rovercode/scripts/commission.py'
```
Instead of working with the RaspberryPi's native python, let's assume that our base image has some version of the Rovercode Docker image on it.

cabarnes commented on services/commissioning.service:
```diff
@@ -0,0 +1,18 @@
+Restart=always
+User=root
+Group=root
+ExecStartPre=-/usr/bin/docker stop rovercode
```
Do you mean to stop `rovercode-commissioning` here?

aninternetof replied:
Yup! Fixing.

aninternetof commented:
I hijacked this PR with the next stage of the work.

I have:
* Pulled the commissioning code out of the `scripts` folder and into its own folder like a proper app
* Added the `.env` setup to the commissioning code
* Added tests for it
* Added its testing and linting to the Travis file

Still to do:
* Work with the RaspPi to get the right /dev file path for the jump drive.
* Add the Wifi setup to the commissioning code

aninternetof commented:
I should probably stop while we review this, so it can merge instead of continuing to grow.

aninternetof commented on README.md:
```diff
@@ -17,38 +17,46 @@ Rovercode runs on a [Raspberry Pi 3](https://www.raspberrypi.org/products/raspbe
-$ sudo docker exec rovercode prospector
+$ docker run --name rovercode-tests -v $PWD:/var/rovercode --entrypoint=/bin/bash rovercode -c 'python -m pytest'
+$ docker run --name rovercode-tests -v $PWD:/var/rovercode --entrypoint=/bin/bash rovercode -c 'prospector'
 
```
I didn't like having to start the service to run the tests.

cabarnes commented on commissioning/README.md:
```diff
@@ -0,0 +1,17 @@
+This is a small service that runs on boot to grab config and Wifi credentials from a thumbdrive.
+
+## Usage
+See how the the `../services/commissioning.service` file uses it.
```
`the the`

aninternetof replied:
Fixing.

cabarnes commented on commissioning/app.py:
```diff
@@ -0,0 +1,55 @@
+    if len(args) != 3:
+        LOGGER.error('Usage: python app.py ENV_SOURCE_DIR ENV_DEST_DIR')
+        return
+    env_source_directory = args[1]
```
You might want to use [argparse](https://docs.python.org/3/howto/argparse.html) for this

aninternetof replied:
I had been scared of mocking it, but I found a good way. Done.

cabarnes commented on commissioning/dest.env:
```diff
@@ -0,0 +1,2 @@
+b=200
```
Is this just example stuff?

aninternetof replied:
Whoops, this was before I successfully mocked out all the file io in the tests. Deleting.

cabarnes commented on commissioning/app.py:
```diff
@@ -0,0 +1,54 @@
+def main():
+    """Kick off all the commissioning work."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument("source", help="source file directory")
```
I think it would be helpful to add validation to these parameters. You could check that the source exists and is readable and that the destination is writable. You can use the [type](https://docs.python.org/3/library/argparse.html#type) parameter to accomplish that

aninternetof replied:
> For example, FileType('w') can be used to create a writable file

So it will actually go ahead and create the file?

cabarnes replied:
I was just thinking that it can throw an error back to the user if the paths are ultimately going to fail instead of just getting a stack trace

aninternetof replied:
Ah, ok.

> type= can take any callable that takes a single string argument and returns the converted value

Do you have any ideas for what that `type=` command would be?

On the source side, I'd like the script to succeed even if the file doesn't exist. In most cases, the Pi will boot with no flashdrive, so the source file won't be present. Right now, the `blob.blob` read gracefully returns an empty list that handles that.

cabarnes commented on services/commissioning.service:
```diff
@@ -0,0 +1,18 @@
+User=root
+Group=root
+ExecStartPre=-/usr/bin/docker stop rovercode-commissioning
+ExecStart=-/usr/bin/docker run --name rovercode-commissioning --privileged -v /dev:/dev -w='/var/rovercode/commissioning' --entrypoint=/bin/bash rovercode -c 'python app.py /dev/something /opt/rovercode/ TODO: use real /dev location'
```
Will the `TODO` be removed later?

aninternetof replied:
Yup. I've got to experiment some on the RaspPi to figure out that argument.

aninternetof replied:
Removed.

jeeshofone commented on services/rovercode.service:
```diff
@@ -10,7 +10,8 @@ User=root
 ExecStartPre=-/usr/bin/docker stop rovercode
 ExecStartPre=-/usr/bin/docker rm rovercode
-ExecStart=-/usr/bin/docker run --name rovercode --privileged -v /dev:/dev -v /opt/rovercode:/var/rovercode rovercode
+ExecStartPre=-/usr/bin/docker pull rovercode/rovercode:latest
```
if you're pulling images there can be a condition where we cannot start because there is not enough free space to pull the new image; we'd need to validate that we can actually pull the image onto disk and also clean up old images

jeeshofone commented on services/rovercode.service:
```diff
@@ -10,7 +10,8 @@ User=root
 ExecStartPre=-/usr/bin/docker stop rovercode
 ExecStartPre=-/usr/bin/docker rm rovercode
-ExecStart=-/usr/bin/docker run --name rovercode --privileged -v /dev:/dev -v /opt/rovercode:/var/rovercode rovercode
+ExecStartPre=-/usr/bin/docker pull rovercode/rovercode:latest
```
if you're pulling images there can be a condition where we cannot start because there is not enough free space to pull the new image; we'd need to validate that we can actually pull the image onto disk and also clean up old images

aninternetof replied:
Good catch. Would the `docker system prune` work in the spot I put it? It is just before removing the old container, so it shouldn't prune the most recent version used by that container.
And, I think this means we'd have at most two images: the one used by the last container, and the new one we're pulling down. Is it safe enough to assume we'll have space for two images?

aninternetof commented on services/commissioning.service:
```diff
@@ -0,0 +1,19 @@
+User=root
+Group=root
+ExecStartPre=-/usr/bin/docker stop rovercode-commissioning
+ExecStartPre=-mkdir -p /mnt/usb-stick && mount /dev/sda1 /mnt/usb-stick
```
Not super robust, but probably enough to get us by in our controlled environment.

aninternetof commented:
I went ahead and added the WiFi setup stuff.
That should be all the features.
